### PR TITLE
Remove bind

### PR DESCRIPTION
**Pull request by peterdavidfagan**
Fixes https://github.com/ros-planning/moveit2/issues/872

This pull request removes std::bind from the MoveIt2 codebase and includes some minor rework of pointer usage. 

The changes in this pull request are based on [these changes](https://github.com/ros-planning/moveit/pull/3106) that were completed within MoveIt (thanks to @v4hn for this).

This is my first pull request on the MoveIt2 project, I hope I didn't overlook anything in the process. I am happy to address any feedback items to ensure this code can be merged into the main project.

### CODE REVIEW

**Comment by peterdavidfagan** [comment]
Hi @henningkayser and @v4hn, 

It looks like I need a maintainer to approve the execution of GHA workflows since I am a first-time contributor. I would appreciate the approval of this workflow run.

I can look to run industrial_ci locally if usage of GHA minutes is an important constraint.

**Comment by peterdavidfagan** [comment]
Thanks for running CI, it looks like one of the tests is failing (`test_ruckig_traj_smoothing`), I will take a look into this to understand why it has occurred.

**Comment by AndyZe** [comment]
Hmm, that's a weird failure. It looks like the test should pass. The requirement is:

(actual_duration < (1.3 * ideal_duration), actual: 0.285312 vs 0.273325

So it's definitely within 30%. I'm confused why that fails

**Comment by AndyZe** [comment]
I'll try running the test again.

**Comment by AndyZe** [comment]
Also, note that the failing test is on `rolling-ci-testing`. That is the upcoming changes that haven't been released yet. The required tests are `colling-ci` and `Formatting` -- both of which pass. So I think you are A-OK  :+1:

**Comment by henningkayser** [comment]
Thank you @peterdavidfagan! The changes look all good at first sight. I would have expected cherry-picked commits from the original PR to keep the original authors of this effort, but I guess we haven't really required that. I'll leave it to @v4hn, @rhaschke to decide if they are ok with the new commits.

**Comment by peterdavidfagan** [comment]
Thanks @henningkayser I appreciate this and I agree it would have been more optimal to preserve the original authors through cherry-picking commits. I would be happy to resubmit this pull request in this format if it is preferred, I would in this case separate any of my changes that are distinct from the original commits.

**Comment by rhaschke** [comment]
Keeping the original authors would have been great indeed. If you don't mind, @peterdavidfagan, please adapt. You can easily change authors via manual rebase and `git commit --amend --author="..."`.

**Comment by peterdavidfagan** [comment]
> Keeping the original authors would have been great indeed. If you don't mind, @peterdavidfagan, please adapt. You can easily change authors via manual rebase and `git commit --amend --author="..."`.

Will do @rhaschke I have set aside time for this at the end of my current work session so I will perform the rebase and push the changes then.

**Comment by peterdavidfagan** [comment]
The above force push resets the authors to those who authored the original commits on the MoveIt repository. For [changes](https://github.com/ros-planning/moveit2/pull/1204/commits/26b93c9fd602ee5e43c4f9ed2eb5a432cd14f1f5) that were made to the planning context manager I assigned @rhaschke as the author since he made the last edits. In this case the commits made to this file have been squashed hence the two sources in the commit description. 

@rhaschke would you prefer for this commit to not be squashed in order to preserve the original history? Happy to also make this change if needed.

**Comment by rhaschke** [comment]
Thanks. Looks good to me.

File: moveit_core/planning_scene/test/test_multi_threaded.cpp
Comment (henningkayser) [inline moveit_core/planning_scene/test/test_multi_threaded.cpp:129]: I've noticed that there are several occasions where pointers are dereferenced like here. Is there a particular reason instead of just dereferencing them in the function call? Just curious @v4hn

**Review by henningkayser — APPROVED** [review]
Ok, diff looks all good to me! CI fails because of an unrelated issue (fixed [here](https://github.com/ros-planning/moveit2/pull/1207)).

File: moveit_core/planning_scene/test/test_multi_threaded.cpp
Comment (rhaschke) [reply moveit_core/planning_scene/test/test_multi_threaded.cpp:129]: I thinks, it's just a matter of what is actually captured, e.g. the full `states` array or a single state (reference).

Comment (henningkayser) [reply moveit_core/planning_scene/test/test_multi_threaded.cpp:129]: Yeah, with states it makes sense to me, with pointers like `planning_scene_` I'm not sure. At least I haven't seen this with local shared pointers afaik, this was my only concern at first.

Comment (v4hn) [reply moveit_core/planning_scene/test/test_multi_threaded.cpp:129]: > Yeah, with states it makes sense to me, with pointers like planning_scene_ I'm not sure. At least I haven't seen this with local shared pointers afaik which was my first concern.

I did that for multiple reasons:
- Restrict scope as Robert explained above
- clang-tidy does it automatically :smiley: 
- Never share ownership with methods that don't need it. It's a rather poisonous pattern in the whole code-base to pass shared pointers for everything, even if no ownership is transferred. It's one of the various things @tylerjw complains about a lot :cat2: 
- Dereferencing pointers when they are captured instead of used makes sure you actually get a traceable error at point of definition and not just at point of use. That pattern can help avoid multiple hours of debugging when `std::function`s are copied around.

**Pull request merged**
